### PR TITLE
[RLlib] Migrate PPO Impala and APPO policies to use sub-classing implementation.

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

This is the last part of the big policy migrate PR.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/impala/vtrace_torch_policy.py [hunk 1]
```diff
@@ -251,51 +163,198 @@ def make_time_major(policy, seq_lens, tensor, drop_last=False):
...
-        "vf_loss": torch.mean(torch.stack(policy.get_tower_stats("vf_loss"))),
-        "vf_explained_var": torch.mean(
-            torch.stack(policy.get_tower_stats("vf_explained_var"))
-        ),
-    }
+    def __init__(self):
+        pass
 
+    @override(TorchPolicyV2)
+    def optimizer(
```
Comment (sven1977) [inline rllib/agents/impala/vtrace_torch_policy.py:173]: Note: We should unify this across all algos. Like why can users chose between Adam and RMSProp for IMPALA, but not for DQN or any other algo?
Doesn't have to be in this PR.

File: rllib/agents/impala/vtrace_torch_policy.py [hunk 2]
```diff
@@ -251,51 +163,198 @@ def make_time_major(policy, seq_lens, tensor, drop_last=False):
...
+                momentum=self.config["momentum"],
+                eps=self.config["epsilon"],
+            )
+
+
+# VTrace mixins are placed in front of more general mixins to make sure
+# their functions like optimizer() overrides all the other implementations
+# (e.g., LearningRateSchedule.optimizer())
+class VTraceTorchPolicy(
+    GradClippingMixin,
```
Comment (sven1977) [inline rllib/agents/impala/vtrace_torch_policy.py:192]: Note (Doesn't have to be in this PR): We should replace mixins - if possible - with method overrides. Most mixins don't have state but only define one method. It would be clearer for the users mental model to simply override the method in the Policy sub-class and then use an imported utility function in there. Mixins are scary, you never know, which one to initialize first to not step over some stuff in the other ones.

File: rllib/agents/impala/vtrace_torch_policy.py [hunk 3]
```diff
@@ -251,51 +163,198 @@ def make_time_major(policy, seq_lens, tensor, drop_last=False):
...
+    ) -> Union[TensorType, List[TensorType]]:
+        """Compute loss for Impala Policy Objective.
+
+        Args:
+            model: The Model to calculate the loss for.
+            dist_class: The action distr. class.
+            train_batch: The training data.
+
+        Returns:
+            The PPO loss tensor given the input batch.
```
Comment (sven1977) [inline rllib/agents/impala/vtrace_torch_policy.py:238]: typo: "IMPALA/VTrace" ?

File: rllib/agents/impala/vtrace_torch_policy.py [hunk 4]
```diff
@@ -251,51 +163,198 @@ def make_time_major(policy, seq_lens, tensor, drop_last=False):
...
-    get_batch_divisibility_req=lambda p: p.config["rollout_fragment_length"],
-)
+    @override(TorchPolicyV2)
+    def loss(
+        self,
+        model: ModelV2,
+        dist_class: Type[ActionDistribution],
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
+        """Compute loss for Impala Policy Objective.
```
Comment (sven1977) [inline rllib/agents/impala/vtrace_torch_policy.py:230]: Actually, do we even need these docstrings? We are overriding a method from a super class. Normally, we don't repeat dosctrings in these cases.

File: rllib/agents/ppo/appo.py [hunk 5]
```diff
@@ -246,11 +245,17 @@ def get_default_policy_class(
         self, config: PartialTrainerConfigDict
     ) -> Optional[Type[Policy]]:
         if config["framework"] == "torch":
-            from ray.rllib.agents.ppo.appo_torch_policy import AsyncPPOTorchPolicy
+            from ray.rllib.agents.ppo.appo_torch_policy import APPOTorchPolicy
```
Comment (sven1977) [inline rllib/agents/ppo/appo.py:248]: Nice!

File: rllib/agents/ppo/appo_tf_policy.py [hunk 6]
```diff
@@ -48,443 +48,425 @@
...
 
 
-def make_appo_model(
-    policy: Policy,
-    obs_space: gym.spaces.Space,
-    action_space: gym.spaces.Space,
-    config: TrainerConfigDict,
-) -> ModelV2:
-    """Builds model and target model for APPO.
+class MakeAPPOModel:
```
Comment (sven1977) [inline rllib/agents/ppo/appo_tf_policy.py:51]: Same here. Let's actually start NOT using mixins wherever possible, but simple method overrides.
Whenever a mixin doesn't have state, we should simply override the mixin's method  in the policy sub-class (here: make_model).

File: rllib/agents/ppo/appo_torch_policy.py [hunk 7]
```diff
@@ -8,391 +8,398 @@
...
-        is_multidiscrete = False
-        output_hidden_shape = 1
-
-    def _make_time_major(*args, **kwargs):
-        return make_time_major(
-            policy, train_batch.get(SampleBatch.SEQ_LENS), *args, **kwargs
+# We need this builder function because we want to share the same
+# custom logics between TF1 dynamic and TF2 eager policies.
+class APPOTorchPolicy(
+    MakeAPPOModel,
```
Comment (sven1977) [inline rllib/agents/ppo/appo_torch_policy.py:60]: Ah, I see: Cross-framework compatibility :) 

Ok, maybe define the utility function behind this in `appo.py` directly? B/c I feel like if it's truly framework agnostic, it shouldn't be in the tf-policy file either.

File: rllib/examples/centralized_critic.py [hunk 8]
```diff
@@ -200,29 +176,47 @@ def central_vf_stats(policy, train_batch, grads):
...
-)
+def get_ccppo_policy(base):
+    class CCPPOTFPolicy(CentralizedValueMixin, base):
+        def __init__(self, observation_space, action_space, config):
+            base.__init__(self, observation_space, action_space, config)
+            CentralizedValueMixin.__init__(self)
+
+        @override(base)
+        def loss(self, model, dist_class, train_batch):
+            return loss_with_central_critic(
```
Comment (sven1977) [inline rllib/examples/centralized_critic.py:187]: Can we add a comment here about the `super()` "trick"? Need to call the super's (actual) loss function inside the central one defined above?

**Review by sven1977 — COMMENTED** [review]
Awesome stuff @gjoliver !
Just some smaller asks and nits. Mostly naming related and then the question on starting to get rid of mixins.
Thanks so much again for wrestling through this tough problem and paving the way for the team!

**Comment by gjoliver** [comment]
ok, let me try to address these comment today.
thanks for the review!

File: rllib/agents/impala/vtrace_torch_policy.py (hunk 1 above)
Comment (gjoliver) [reply rllib/agents/impala/vtrace_torch_policy.py:173]: ok, added a TODO.
this may be agent / model dependent I think?

File: rllib/agents/impala/vtrace_torch_policy.py (hunk 2 above)
Comment (gjoliver) [reply rllib/agents/impala/vtrace_torch_policy.py:192]: done.
I left compute_gradients() in tf_mixins.py, which tbh may not be the best place.
added a TODO about it.

Comment (gjoliver) [reply rllib/agents/impala/vtrace_torch_policy.py:192]: otherwise, GAE and ComputeAndClipGrad are not MixIns anynore.

File: rllib/agents/impala/vtrace_torch_policy.py (hunk 3 above)
Comment (gjoliver) [reply rllib/agents/impala/vtrace_torch_policy.py:238]: done

File: rllib/agents/impala/vtrace_torch_policy.py (hunk 4 above)
Comment (gjoliver) [reply rllib/agents/impala/vtrace_torch_policy.py:230]: yeah, I don't think we need them. ok, getting rid of as much as I can find.

File: rllib/agents/ppo/appo_tf_policy.py (hunk 6 above)
Comment (gjoliver) [reply rllib/agents/ppo/appo_tf_policy.py:51]: done

File: rllib/agents/ppo/appo_torch_policy.py (hunk 7 above)
Comment (gjoliver) [reply rllib/agents/ppo/appo_torch_policy.py:60]: I am keeping this in appo_tf_policy.py, since policy shouldn't import trainer.py either.
maybe add an appo_util.py or something ...

File: rllib/examples/centralized_critic.py (hunk 8 above)
Comment (gjoliver) [reply rllib/examples/centralized_critic.py:191]: yeah, that's right.
added a comment.

File: rllib/tests/test_execution.py [hunk 9]
```diff
@@ -5,7 +5,7 @@
 import unittest
 
 import ray
-from ray.rllib.agents.ppo.ppo_tf_policy import PPOTFPolicy
+from ray.rllib.agents.ppo.ppo_tf_policy import PPODynamicTFPolicy
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/tests/test_execution.py]: done.
are you ok with XXStaticGraphTFPolicy and XXEagerTFPolicy?

**Review by gjoliver — COMMENTED** [review]
ok, I think I addressed all the comments.
let's wait for the tests to pass.

File: rllib/agents/impala/vtrace_tf_policy.py [hunk 10]
```diff
@@ -158,197 +167,273 @@ def _make_time_major(policy, seq_lens, tensor, drop_last=False):
...
+            VTraceClipGradients.__init__(self)
+            VTraceOptimizer.__init__(self)
+            LearningRateSchedule.__init__(self, config["lr"], config["lr_schedule"])
+            EntropyCoeffSchedule.__init__(
+                self, config["entropy_coeff"], config["entropy_coeff_schedule"]
+            )
 
+            # Note: this is a bit ugly, but loss and optimizer initialization must
+            # happen after all the MixIns are initialized.
+            self.maybe_initialize_optimizer_and_loss()
```
Comment (avnishn) [inline rllib/agents/impala/vtrace_tf_policy.py:307]: is there a reason why this is called maybe... does that imply that it doesnt happen always, and if so, why?

**Review by avnishn — APPROVED** [review]

File: rllib/agents/impala/vtrace_tf_policy.py (hunk 10 above)
Comment (gjoliver) [reply rllib/agents/impala/vtrace_tf_policy.py:307]: TF2 doesn't need to initialize these. You only need to construct loss graphs for static graph TF1.
TF2 takes the chance to initialize a few things like view_requirements I think.

File: rllib/agents/ppo/appo_torch_policy.py (hunk 7 above)
Comment (sven1977) [reply rllib/agents/ppo/appo_torch_policy.py]: Hmhm, importing the trainer module from policy works if we do the import in-line.

File: rllib/tests/test_execution.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/tests/test_execution.py]: YES! Thanks!
